server/datastore/s3/carves.go

```diff
@@ -0,0 +1,236 @@
+package s3
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/fleetdm/fleet/server/kolide"
+	"github.com/pkg/errors"
+)
+
+const (
+	defaultMaxS3Keys = 1000
+	cleanupSize      = 1000
+	// This is Golang's way of formatting timestrings, it's confusing, I know.
+	// If you are used to more conventional timestrings, this is equivalent
+	// to %Y/%m/%d/%H (year/month/day/hour)
+	timePrefixFormat = "2006/01/02/15"
+)
+
+// generateS3Key builds S3 key from carve metadata
+// all keys are prefixed by date so that they can easily be listed chronologically
+func (d *Datastore) generateS3Key(metadata *kolide.CarveMetadata) string {
+	simpleDateHour := metadata.CreatedAt.Format(timePrefixFormat)
+	return fmt.Sprintf("%s%s/%s", d.prefix, simpleDateHour, metadata.Name)
+}
+
+// NewCarve initializes a new file carving session
+func (d *Datastore) NewCarve(metadata *kolide.CarveMetadata) (*kolide.CarveMetadata, error) {
+	objectKey := d.generateS3Key(metadata)
+	res, err := d.s3client.CreateMultipartUpload(&s3.CreateMultipartUploadInput{
+		Bucket: &d.bucket,
+		Key:    &objectKey,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "s3 multipart carve create")
+	}
+	metadata.SessionId = *res.UploadId
+	return d.metadatadb.NewCarve(metadata)
+}
+
+// UpdateCarve updates carve definition in database
+// Only max_block and expired are updatable
+func (d *Datastore) UpdateCarve(metadata *kolide.CarveMetadata) error {
+	return d.metadatadb.UpdateCarve(metadata)
+}
+
+// listS3Carves lists all keys up to a given one or if the passed max number
+// of keys has been reached; keys are returned in a set-like map
+func (d *Datastore) listS3Carves(lastPrefix string, maxKeys int) (map[string]bool, error) {
+	var err error
+	var continuationToken string
+	result := make(map[string]bool)
+	if maxKeys <= 0 {
+		maxKeys = defaultMaxS3Keys
+	}
+	if !strings.HasPrefix(lastPrefix, d.prefix) {
+		lastPrefix = d.prefix + lastPrefix
+	}
+	for {
+		carveFilesPage, err := d.s3client.ListObjectsV2(&s3.ListObjectsV2Input{
+			Bucket:            &d.bucket,
+			Prefix:            &d.prefix,
+			ContinuationToken: &continuationToken,
+		})
+		if err != nil {
+			return nil, err
+		}
+		for _, carveObject := range carveFilesPage.Contents {
+			result[*carveObject.Key] = true
```

Comment:
I think `*` is unnecessary in Go here and many other places below. I will try to comment them all.
```suggestion
			result[carveObject.Key] = true
```

Reply:
Mmm, I don't know about this. The AWS sdk returns everything as a reference and my linter was getting pretty angry if I didn't properly de-reference them. I'll take a close look to be sure.

Reply:
I confirm that de-referencing is very much required